editor/plugins/text_control_editor_plugin.cpp

```diff
@@ -121,78 +122,162 @@ void TextControlEditor::_update_styles_menu() {
 			} else {
-				// Single face font, search for the font with matching name and style.
-				String name = font->get_data(0)->get_font_name();
-				String style = font->get_data(0)->get_font_style_name();
-				if (fonts.has(name) && fonts[name].has(style)) {
-					_update_fonts_menu();
-					for (int i = 0; i < font_list->get_item_count(); i++) {
-						if (font_list->get_item_text(i) == name) {
-							font_list->select(i);
-							break;
-						}
+				edited_color = "font_color";
+				edited_font = "font";
+				edited_font_size = "font_size";
+			}
+
+			// Get font override.
+			Ref<Font> font;
+			if (edited_control->has_theme_font_override(edited_font)) {
+				font = edited_control->get_theme_font(edited_font);
+			}
+
+			if (font.is_valid()) {
+				if (font->get_data_count() != 1) {
+					custom_font = font;
+					if (i > 0) {
+						same_font = same_font && (font_selected == FONT_INFO_USER_CUSTOM);
+						same_style = same_style && (style_selected == 0);
 					}
-					_update_styles_menu();
-					for (int i = 0; i < font_style_list->get_item_count(); i++) {
-						if (font_style_list->get_item_text(i) == style) {
-							font_style_list->select(i);
-							break;
+
+					font_selected = FONT_INFO_USER_CUSTOM;
+					style_selected = 0;
+				} else {
+					String name = font->get_data(0)->get_font_name();
+					String style = font->get_data(0)->get_font_style_name();
+					if (fonts.has(name) && fonts[name].has(style)) {
+						for (int j = 0; j < font_list->get_item_count(); j++) {
+							if (font_list->get_item_text(j) == name) {
+								if (i > 0) {
+									same_font = same_font && (j == font_selected);
+								}
+
+								font_selected = j;
+								break;
+							}
 						}
+						for (int j = 0; j < font_style_list->get_item_count(); j++) {
+							if (font_style_list->get_item_text(j) == style) {
+								if (i > 0) {
+									same_style = same_style && (j == style_selected);
+								}
+
+								style_selected = j;
+								break;
+							}
+						}
+					} else {
+						custom_font = font;
+						if (i > 0) {
+							same_font = same_font && (font_selected == FONT_INFO_USER_CUSTOM);
+							same_style = same_style && (style_selected == 0);
+						}
+
+						font_selected = FONT_INFO_USER_CUSTOM;
+						style_selected = 0;
 					}
-				} else {
-					// Unknown font, save it to "custom_font" to allow undoing font change.
-					custom_font = font;
-					_update_fonts_menu();
-					font_list->select(FONT_INFO_USER_CUSTOM);
-					_update_styles_menu();
-					font_style_list->select(0);
 				}
+			} else {
+				if (i > 0) {
+					same_font = same_font && (font_selected == FONT_INFO_THEME_DEFAULT);
+					same_style = same_style && (style_selected == 0);
+				}
+
+				font_selected = FONT_INFO_THEME_DEFAULT;
+				style_selected = 0;
 			}
+
+			int current_font_size = edited_control->get_theme_font_size(edited_font_size);
+			int current_outline_size = edited_control->get_theme_constant("outline_size");
```

Comment:
This (and line 225) can wrap the string literal in `SNAME` to get a statically allocated StringName instead of an implicit coercion.